src/main/java/org/photonvision/mrcal/MrCalJNI.java

```diff
@@ -62,33 +62,49 @@ public static native MrCalResult mrcal_calibrate_camera(
             int boardWidth, int boardHeight, double boardSpacing,
             int imageWidth, int imageHeight, double focalLen);
 
-    public static native boolean undistort_mrcal(long srcMat, long dstMat, long cameraMat, long distCoeffsMat, int lensModelOrdinal, int order, int Nx, int Ny, int fov_x_deg);
+    public static native boolean undistort_mrcal(long srcMat, long dstMat, long cameraMat, long distCoeffsMat,
+            int lensModelOrdinal, int order, int Nx, int Ny, int fov_x_deg);
 
-    public static MrCalResult calibrateCamera(
-            List<MatOfPoint2f> board_corners,
-            int boardWidth, int boardHeight, double boardSpacing,
-            int imageWidth, int imageHeight, double focalLen) {
-        double[] observations = new double[boardWidth * boardHeight * 3 * board_corners.size()];
+            public static MrCalResult calibrateCamera(
+                List<MatOfPoint2f> board_corners,
+                List<MatOfFloat> board_corner_levels,
+                int boardWidth, int boardHeight, double boardSpacing,
+                int imageWidth, int imageHeight, double focalLen) {
+            double[] observations = new double[boardWidth * boardHeight * 3 * board_corners.size()];
+
+            if (board_corners.size() != board_corner_levels.size()) {
+                return new MrCalResult(false);
+            }
 
-        int i = 0;
-        for (var board : board_corners) {
-            var corners = board.toArray();
-            // Assume that we're correct in terms of row/column major-ness (lol)
-            for (var c : corners) {
-                float level = 1.0f; // if we have mrgingham, use level from that. Otherwise, hard-coded to 1
+            int i = 0;
+            for (int b = 0; b < board_corners.size(); b++) {
+                var board = board_corners.get(b);
+                var levels = board_corner_levels.get(b).toArray();
+                var corners = board.toArray();
+                // Assume that we're correct in terms of row/column major-ness (lol)
+                for (int c = 0; c < corners.length; c++) {
+                    if (corners.length != levels.length) {
```

Comment:
or wait you check the length here, but never that the length is actually correct? we should do that.